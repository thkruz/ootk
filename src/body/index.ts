export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
