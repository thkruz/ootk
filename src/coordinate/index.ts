export * from './Hill.js';
export * from './ClassicalElements.js';
export * from './EquinoctialElements.js';
export * from './FormatTle.js';
export * from './Geodetic.js';
export * from './ITRF.js';
export * from './J2000.js';
export * from './RelativeState.js';
export * from './RIC.js';
export * from './StateVector.js';
export * from './TEME.js';
export * from './Tle.js';
