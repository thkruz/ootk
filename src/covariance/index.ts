export * from './CovarianceSample.js';
export * from './StateCovariance.js';
