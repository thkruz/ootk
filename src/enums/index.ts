export { CatalogSource } from './CatalogSource.js';
export { CommLink } from './CommLink.js';
export { OrbitRegime } from './OrbitRegime.js';
export { Sgp4OpsMode } from './Sgp4OpsMode.js';
export { AngularDiameterMethod } from './AngularDiameterMethod.js';
export { AngularDistanceMethod } from './AngularDistanceMethod.js';
export { PassType } from './PassType.js';
