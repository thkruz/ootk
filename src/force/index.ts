export { EarthGravity } from './EarthGravity.js';
export { Force } from './Force.js';
export { ForceModel } from './ForceModel.js';
export { Gravity } from './Gravity.js';
export { SolarRadiationPressure } from './SolarRadiationPressure.js';
export { ThirdBodyGravity } from './ThirdBodyGravity.js';
export { Thrust } from './Thrust.js';

