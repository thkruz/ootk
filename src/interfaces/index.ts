export type { GroundPositionParams } from './GroundPositionParams.js';
export type { OptionsParams } from './OptionsParams.js';
export type { TimeVariables } from './TimeVariables.js';
export type { StarObjectParams } from './StarObjectParams.js';
export type { ClassicalElementsParams } from './ClassicalElementsParams.js';
export type { EquinoctialElementsParams } from './EquinoctialElementsParams.js';
export type { ScenarioParams } from '../objects/Scenario.js';
export type { BaseObjectParams } from '../objects/base/BaseObject.js';
export type { GroundObjectParams } from '../objects/base/GroundObject.js';
export type { FacilityParams } from '../objects/base/Facility.js';
export type { SensorParams } from '../objects/attachable/SensorAdv.js';
export type { RectangularSensorParams } from '../objects/attachable/RectangularSensor.js';
