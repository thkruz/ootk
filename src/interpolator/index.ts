export { ChebyshevCoefficients } from './ChebyshevCoefficients.js';
export { ChebyshevInterpolator } from './ChebyshevInterpolator.js';
export { CubicSpline } from './CubicSpline.js';
export { CubicSplineInterpolator } from './CubicSplineInterpolator.js';
export { Interpolator } from './Interpolator.js';
export { LagrangeInterpolator } from './LagrangeInterpolator.js';
export { StateInterpolator } from './StateInterpolator.js';
export { VerletBlendInterpolator } from './VerletBlendInterpolator.js';

