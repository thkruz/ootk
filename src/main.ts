/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
