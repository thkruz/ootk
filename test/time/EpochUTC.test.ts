import { EpochUTC, Seconds } from '../../src/main';

describe('EpochGPS', () => {
  it('now', () => {
    const epoch = EpochUTC.now();
    const now = new Date();

    expect(epoch.posix).toBeCloseTo(now.getTime() / 1000, 2);
  });

  it('fromDate', () => {
    const epoch = EpochUTC.fromDate({ year: 2021, month: 3, day: 1 });

    expect(epoch).toMatchSnapshot();
  });

  it('fromDateTime', () => {
    const epoch = EpochUTC.fromDateTime(new Date(2021, 2, 1));

    expect(epoch).toMatchSnapshot();
  });

  it('fromDateTimeString', () => {
    const epoch = EpochUTC.fromDateTimeString('2021-03-01');

    expect(epoch).toMatchSnapshot();
  });

  it('fromJ2000TTSeconds', () => {
    const epoch = EpochUTC.fromJ2000TTSeconds(0 as Seconds);

    expect(epoch).toMatchSnapshot();
  });

  it('fromDefinitiveString', () => {
    const epoch = EpochUTC.fromDefinitiveString('1/2021 00:00:00');

    expect(epoch).toMatchSnapshot();
  });

  // roll
  it('should roll the epoch forward', () => {
    const epoch = EpochUTC.fromDateTime(new Date(2021, 2, 1));

    expect(epoch.roll(60 as Seconds)).toMatchSnapshot();
  });

  // toMjd
  it('should convert to an MJD', () => {
    const epoch = EpochUTC.fromDateTime(new Date(2021, 2, 1));

    expect(epoch.toMjd()).toMatchSnapshot();
  });

  // toMjdGsfc
  it('should convert to an MJD GSFC', () => {
    const epoch = EpochUTC.fromDateTime(new Date(2021, 2, 1));

    expect(epoch.toMjdGsfc()).toMatchSnapshot();
  });

  // toTAI
  it('should convert to a TAI epoch', () => {
    const epoch = EpochUTC.fromDateTime(new Date(2021, 2, 1));

    expect(epoch.toTAI()).toMatchSnapshot();
  });

  // toTT
  it('should convert to a TT epoch', () => {
    const epoch = EpochUTC.fromDateTime(new Date(2021, 2, 1));

    expect(epoch.toTT()).toMatchSnapshot();
  });

  // toTDB
  it('should convert to a TDB epoch', () => {
    const epoch = EpochUTC.fromDateTime(new Date(2021, 2, 1));

    expect(epoch.toTDB()).toMatchSnapshot();
  });

  // toGPS
  it('should convert to a GPS epoch', () => {
    const epoch = EpochUTC.fromDateTime(new Date(2021, 2, 1));

    expect(epoch.toGPS()).toMatchSnapshot();
  });

  // gmstAngle
  it('should calculate the GMST angle', () => {
    const epoch = EpochUTC.fromDateTime(new Date(2021, 2, 1));

    expect(epoch.gmstAngle()).toMatchSnapshot();
  });

  // gmstAngleDegrees
  it('should calculate the GMST angle in degrees', () => {
    const epoch = EpochUTC.fromDateTime(new Date(2021, 2, 1));

    expect(epoch.gmstAngleDegrees()).toMatchSnapshot();
  });
});
