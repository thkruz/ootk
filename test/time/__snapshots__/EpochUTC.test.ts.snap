// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpochGPS fromDate 1`] = `
EpochUTC {
  "posix": 1614556800,
}
`;

exports[`EpochGPS fromDateTime 1`] = `
EpochUTC {
  "posix": 1614574800,
}
`;

exports[`EpochGPS fromDateTimeString 1`] = `
EpochUTC {
  "posix": 1614556800,
}
`;

exports[`EpochGPS fromDefinitiveString 1`] = `
EpochUTC {
  "posix": 1609459200,
}
`;

exports[`EpochGPS fromJ2000TTSeconds 1`] = `
EpochUTC {
  "posix": 946727935.816,
}
`;

exports[`EpochGPS should calculate the GMST angle 1`] = `4.088037525758033`;

exports[`EpochGPS should calculate the GMST angle in degrees 1`] = `234.22729671703883`;

exports[`EpochGPS should convert to a GPS epoch 1`] = `
EpochGPS {
  "seconds": 104417.99999993818,
  "week": 2147,
}
`;

exports[`EpochGPS should convert to a TAI epoch 1`] = `
EpochTAI {
  "posix": 1614574837,
}
`;

exports[`EpochGPS should convert to a TDB epoch 1`] = `
EpochTDB {
  "posix": 1614574869.1853862,
}
`;

exports[`EpochGPS should convert to a TT epoch 1`] = `
EpochTT {
  "posix": 1614574869.184,
}
`;

exports[`EpochGPS should convert to an MJD 1`] = `59274.20833333349`;

exports[`EpochGPS should convert to an MJD GSFC 1`] = `29274.70833333349`;

exports[`EpochGPS should roll the epoch forward 1`] = `
EpochUTC {
  "posix": 1614574860,
}
`;
